### PR TITLE
Feature/14 supervised fine tuning

### PR DESCRIPTION
**Pull request by erybie222**
This PR introduces a complete Supervised Fine-Tuning (SFT) pipeline for the `Bielik-11B-v3.0-Instruct` model. It uses **QLoRA** for memory efficiency and is fully configured for HPC clusters using SLURM.

### CODE REVIEW

**Comment by Igor-S-666** [comment]
Ogółem kod wygląda ok, niemniej mam troche uwag (na razie go nie testowałem na H100, jutro to zrobię i dam znać czy jakieś inne błędy wykryłem):


## Major Problems

- Wrong chat template for Bielik: in `train.py:11–19`, the
`formatting_prompts_func` function uses the **Llama-2 chat format** 
```python
return f"<s>[INST] {instruction} [/INST] {response} </s>"
```
Yet Bielik-11B-v3.0-Instruct uses **ChatML** as mentioned on official huggingface card. Fine-tuning on the wrong template teaches the model a format it was never pretrained on. Possibly degrading performance. To fix, the function should use the tokenizer's own `apply_chat_template` method, which applies the correct ChatML format for Bielik. Read more here: https://huggingface.co/learn/llm-course/en/chapter11/2 \
And there: https://huggingface.co/speakleash/Bielik-11B-v3.0-Instruct section "chat template".
Here proposition of fix (warning: it was AI generated, do not guarantee it is correct, but it should be close):

```python
def formatting_prompts_func(example):
    messages = [
        {"role": "user",    "content": example["query"]},
        {"role": "assistant","content": example["response"]},
    ]
    return tokenizer.apply_chat_template(messages, tokenize=False, add_generation_prompt=False)
```

---

- No `torch_dtype` when loading without quantization (`train.py:71–75`)

```python
base_model = AutoModelForCausalLM.from_pretrained(
    cfg.BASE_MODEL,
    device_map="auto",
    quantization_config=quant_config
)
```

Without `torch_dtype`, `transformers` defaults to **float32**, which doubles memory usage but makes no sense, as huggingface card clearly states that bielik-11b-v3-instruct is available in **bfloat16**. 
To fix, add a simple default dtype.


## Minor Problems

- No random seed setted across the code, especially in validation split(`train.py:37`)
```python
dataset = raw_dataset.train_test_split(test_size=cfg.VAL_SIZE)
```

Without a `seed=` argument, each run draws a different validation split. It makes our experiments non-reproducible and can lead to noisy eval results. TO fix, set a fixed seed for the split.

Also, consider setting a global seed at the start of `train.py` on all non-deterministic libraries (like torch, numpy, trensformers) for full reproducibility.

- In `train.py:67` you do:

```python
tokenizer.pad_token = tokenizer.eos_token
```

Which seems a bad design choice; pad tokens act as a neutral "no input" signal, which normalizes text length, while the EOS token is a strong "end of text" signal. To fix, use a dedicated pad token instead of the EOS token. 

- Using `gradient_checkpointing=True` without `use_reentrant=False` (`train.py:120`)

In newer Pytorch/transformers versions, you must set up use_reentrant parameter. It is highly advised to set `use_reentrant=False` when using gradient checkpointing with large models, as  the default `use_reentrant=True` can cause unexpected OOM errors on large models. To fix, add `gradient_checkpointing_kwargs={"use_reentrant": False}` to the `SFTConfig`.

- Move `save_total_limit` and `per_device_eval_batch_size=1` to Config; Generally it's a bad practice to have hardcoded values ("Magic Numbers") in the code. \
Also, batch_size of 1 for evaluation is very slow, especially on H100. Consider increasing it to 4 or 8 by default.


## Potential Improvements

- How did you choose cuda version in `requirements.txt`? (`torch==2.3.0+cu118`). While it is minimally functional on H100, maybe worth switching to `cu121` or `cu124` for better performance and compatibility with sm_90 features (if possible of course).

- Formatting function processes examples one at a time (`train.py:11`): `formatting_prompts_func(example)` receives a single dict. `SFTTrainer` supports batched formatting functions (`example` is a dict of lists), which avoids Python function call overhead per sample.

- `BATCH_SIZE = 16` will OOM in the no-quantization path (`config.py:35`). It would be good idea to change default setting to sth plausible on H100. To fix,  Set a safe default — `BATCH_SIZE = 2` or `4` or enable quantization by default (`QUANTIZATION = "4b"`).

- Load best checkpoint at end of training (`train.py:100`) to ensure the best checkpoint, not the last checkpoint, is used for evaluation and saved as the final model.

- `paged_adamw_8bit` for quantized runs: When running with 8-bit quantization, `paged_adamw_8bit` is more memory-efficient than `paged_adamw_32bit`. Make the optimizer choice dependent on the quantization mode.

- use flash attention (version 2 or 3): On H100, enabling Flash Attention e.g.(`attn_implementation="flash_attention_2"` in `AutoModelForCausalLM.from_pretrained` aka `train.py:71`) provides a large reduction in memory and a significant speedup during training.
For general Idea of flash attention, read here: https://medium.com/better-ml/the-free-lunch-of-flash-attention-036b0040dee2

- Validate `QUANTIZATION` value early: Make simple assertion at the start of `train.py` to ensure `QUANTIZATION` is set to a valid value. e.g.

```python
assert cfg.QUANTIZATION in ("none", "4b", "8b"), f"Unknown QUANTIZATION: {cfg.QUANTIZATION}"
```

- Use Dataclass for Config class in `config.py`: Instead of a plain class, use a `@dataclass` for `Config` in `config.py`, as this class currently stores only data, and that's suitable structure for it, not regular class.

**Comment by Igor-S-666** [comment]
W ogóle sorry że dałem review jako komentarz a nie za pomocą add review; jakoś nie zauważyłem tego okienka, dopiero teraz to widzę; na razie zostanę tu bo nie chce chaosu wprowadzać;

Przypomniało mi się że w setup.sh i run_training.sh hardcodeujesz nr grantu projektu, więc każdy kto chce skryptu używać musi to ręcznie zmienić; Albo wrzuć to jako pole w .env albo co najmniej oznacz to w readme.md

**Comment by erybie222** [comment]
### Fixed problems:

* Now formatting function uses ChatML format (via `tokenizer.apply_chat_template`) to match Bielik's pretraining.
* Loading model - now model loads in adequate float precision to prevent memory doubling.
* Random seed - now in config there is set global seed  which is used in dataset loading, SFT config, model, and tokenizer loading for full reproducibility.
* `use_reentrant` parameter set to `False` in gradient checkpointing kwargs to prevent OOM errors.
* Moved `save_total_limit` and `per_device_eval_batch_size` to `Config` to avoid magic numbers.
* Changed evaluation batch_size (increased for H100 efficiency).
* Pad token issue fixed: fallback to `<unk>` token instead of repurposing the `<eos>` token, maintaining semantic correctness.

### Improvements:

* Use newer cuda version (cu124) for better compatibility with H100 (sm_90).
* Set safe train batch size to 4 (prevents OOM on non-quantized runs).
* Save best checkpoint at the end of training (`load_best_model_at_end=True`).
* Use adequate optimizer version - now it depends on quantization settings (e.g., `paged_adamw_8bit` when 8-bit quantization is active).
* Validate quantization env variable early at the beginning of training with an `assert`.
* Refactored `Config` class using `@dataclass` for cleaner data structure.
* Grant number is fetched by scripts from env file 

### Addressed specific review points (Why some things were done differently):

* **Flash Attention and SDPA:** Instead of adding `attn_implementation="flash_attention_2"` and the external `flash-attn` dependency, I decided on a different solution: **`attn_implementation="sdpa"`**. Since PyTorch 2.0+, the built-in Scaled Dot Product Attention on H100 cards with `bf16` automatically invokes optimized FlashAttention-2 kernels underneath. Manually setting FlashAttention-2 caused  errors on load  related to packages
* **Batch mode in formatting_func:** Despite optimization suggestions, I left the single-line processing. In newer versions of the TRL library, the `SFTTrainer` architecture has changed. The formatting function receives a single row by default and must return the `str` type (not `list[str]`). Attempting to implement batch mode with the new TRL resulted in Python iterating over the characters of a single string, putting the shredded letters into the "text" column.

**Comment by Igor-S-666** [comment]
Ogółem dobrze, fajnie że poprawiłeś błędy i mam nadzieje że będzie szybkie; niemniej pojawiło się kilka drobnych błędów do poprawy:

### Minor problems:

- dataclasses if/else in `config.py:64-69` does not work properly, you should use `__post_init__` method and there put your if/else clause; see https://stackoverflow.com/questions/67432117/if-else-in-python-dataclasses
- in `config.py:77` you put a comma: `save_total_limit = 10,` which makes it a tuple, not int as intended; remove the comma (I feel really bad writing this as an issue)
- in `config.py:10:11` you assume .env file is in the root (parents[1] so root of whole repo), but in the instructions you say to run the setup and training scripts from this folder directory.
- in `train.py:17` you use tokenizer, but it was never defined for now; to fix it, you may claim it as argument or a global variable.
- in `train.py:77` in `AutoModelForCausalLM.from_pretrained` you have dtype argument; AI says it does not exist and will raise an error, and strangely huggingface documentation rejects to cooperate with me; could you verify that?

### Potential improvements:
- Pytorch is imported twice: both by installing requirements.txt and from setup.sh
- Readme says the default QUANTIZATION is none, but in .env.example it's set to "4b"; 
- I forgot to mention before, but when I tried to use thus script I got unexpected crash because I gave wrong HF token; Maybe it's good idea to add try/except blocks in crucial points: 
  - when loading the tokenizer and model from HuggingFace (in case of wrong token or model name)
  - when loading the dataset (in case of wrong path or format)
  - when saving checkpoints (in case of disk issues)
- remove line `train.py:69` aka `tokenizer.pad_token = tokenizer.eos_token`; it is not used making it misleading.

**Comment by erybie222** [comment]
### Fixed:
- Dataclass conditionals handling moved to __post_init__ magic method and use fields
- Got rid of `config.py:77` `save_total_limit = 10,` which was error while typing
- Fixed env path in config - now it indicates inside module instead of root
- Got rid of  ` tokenizer.pad_token = tokenizer.eos_token` , it was left in code by mistake
- Got rid of doubled Pytorch installation
- 
### Improvements:
- Add try/except blocks in crucial points
- Refactored train script into Trainer class 

### Comments:
- Hugging Face AutoModelForCausalLM.from_pretrained function uses dtype parameter in newer versions, when I used torch_dtype It gave alert that the parameter name should be changed because it is obsolete https://huggingface.co/docs/transformers/v5.10.1/en/main_classes/model#transformers.PreTrainedModel.from_pretrained

**Comment by erybie222** [comment]
I tested the module on H100 cluster after the changes and the training run was succesful.

**Comment by Igor-S-666** [comment]
Ok, it seems everything is fine now, All issues fixed
We can merge now
